src/libserver/milter.c

```diff
@@ -1781,13 +1781,13 @@ rspamd_milter_process_milter_block (struct rspamd_milter_session *session,
 			it = NULL;
 
 			while ((cur = ucl_object_iterate (elt, &it, true)) != NULL) {
-				ucl_object_iter_t *elt_it;
+				const char *key;
 
-				elt_it = ucl_object_iterate_new (cur);
+				key = ucl_object_key (cur);
 
-				while ((cur_elt = ucl_object_iterate_safe (elt_it, true)) != NULL) {
+				if (key != NULL && (cur_elt = ucl_object_lookup(elt, key)) != NULL) {
```

Comment:
I'm sorry, but your change is totally not clear to me. Initially, the idea was to perform iteration on each header, as it can have multiple value. What you do by `cur_elt = ucl_object_lookup(elt, key)` is efficiently the same as assigning `cur = cur_elt` and nothing else.

Reply:
I'm not familiar with libucl, but as far as I understood that, if `elt` is a hash, `ucl_object_iterate()` returns the result of [`ucl_hash_iterate2()`](https://github.com/rspamd/rspamd/blob/3.2/contrib/libucl/ucl_hash.c#L310):

https://github.com/rspamd/rspamd/blob/c1b767c23463735b645725419adbf7ef60c7e948/contrib/libucl/ucl_util.c#L2715-L2729

It is _not_ the same as the object we want _but_ each hash element (`ucl_hash_elt *`) in `elt`.  That's why I executed `ucl_object_lookup()` to get `cur_elt`.

Second, I don't understand why the iteration is doubled in the current code.  you say "_it can have multiple value_", however, the multi-value case looks made by handling the case that the type of `cur_elt` is `UCL_ARRAY` (Conversely, since the current code handles the `UCL_ARRAY` case in the second iteration, IMHO the third iteration to process `UCL_ARRAY` is never executed).

Iterating on the `UCL_OBJECT` case makes no sense.  In fact, current code treats a hash like `{value = 'header value', order = 1}` as "multiple value" `'header value'` and `1` (latter is ignored because it is not UC_STRING).

Reply:
> Iterating on the `UCL_OBJECT` case makes no sense. In fact, current code treats a hash like `{value = 'header value', order = 1}` as "multiple value" `'header value'` and `1` (latter is ignored because it is not `UCL_STRING`).

To confirm above, modify `milter_headers.lua` as:
``` diff
--- a/src/plugins/lua/milter_headers.lua
+++ b/src/plugins/lua/milter_headers.lua
@@ -528,7 +528,7 @@ local function milter_headers(task)
             settings.routines['authentication-results']))

     if res then
-      add_header('authentication-results', res, ';', 1)
+      add_header('authentication-results', res, ';', '1')
     end
   end
```
then send a message.  As the result with current code, header modified by rspamd is like:
``` code
...

MIME-Version: 1.0
Content-Language: en-US
Content-Type: text/plain; charset=UTF-8; format=flowed
Content-Transfer-Encoding: 7bit
Authentication-Results: 1
Authentication-Results: example.co.jp;
        none

body
```
Because `add_header()` generates a hash `{value = 'header value', order = '1'}` for `Authentication-Results`. As the type of `'1'` is `UCL_STRING`, it is not ignored.

(Perhaps it is also possible to set `default_headers_order = "1"` (a string), but I haven't tried it.)

Reply:
Ok, I probably see what do you mean, and there is another solution: `ucl_object_iterate_safe` accepts an argument `expand_values` that either iterate over multiple values (when set to `false`) or over each individual element (when set to `true`). So, in this particular case, the sollution might be just replacing `true` to `false` as it was intended by the logic of the code.

Reply:
Well, it won't really help, as there is a logical mistake in the code (or in the documentation). Let me think about other possibilities...